### PR TITLE
test: make sure that reading a directory fails

### PR DESCRIPTION
**Pull request by thefourtheye**
Fixes: https://github.com/libuv/libuv/issues/2026

This is the test from #2023 with the changes requested by @santigimeno.

cc @bnoordhuis

### CODE REVIEW

File: test/test-fs.c [hunk 1]
```diff
@@ -3037,6 +3037,54 @@ TEST_IMPL(fs_write_alotof_bufs_with_offset) {
...
+  /* Setup Done Here */
+
+  /* Get a file descriptor for the directory */
+  r = uv_fs_open(loop,
+                 &open_req1,
+                 "test_dir",
+                 O_RDONLY | O_DIRECTORY,
+                 S_IWUSR | S_IRUSR,
+                 NULL);
+  ASSERT(r > 0);
```
Comment (santigimeno) [inline test/test-fs.c]: nit: `ASSERT(r >= 0);`

File: test/test-fs.c [hunk 2]
```diff
@@ -3037,6 +3037,54 @@ TEST_IMPL(fs_write_alotof_bufs_with_offset) {
...
+  /* Try to read data from the directory */
+  iov = uv_buf_init(buf, sizeof(buf));
+  r = uv_fs_read(NULL, &read_req, open_req1.result, &iov, 1, 0, NULL);
+#if defined(__FreeBSD__) || \
+    defined(__OpenBSD__) || \
+    defined(__NetBSD__)  || \
+    defined(_BSD_SOURCE) || \
+    defined(_AIX)
+  /* As of now, only BSDs and AIX support reading from a directory. */
+  ASSERT(r == 0)
```
Comment (santigimeno) [inline test/test-fs.c]: I would change this to `ASSERT((r >= 0) || r == UV_EISDIR)` as the result depends on the filesystem where the directory lives. I would also add `smartos` here too (see: http://illumos.org/man/read.2) and also `DragonFly`.

Comment (thefourtheye) [reply test/test-fs.c]: I included `__sun` and `__DragonFly__`, also made the condition the assertion a bit lenient as you suggested.

**Comment by thefourtheye** [comment]
@santigimeno @carenas I updated the test as per your suggestions, PTAL.

File: test/test-fs.c [hunk 3]
```diff
@@ -3037,6 +3037,60 @@ TEST_IMPL(fs_write_alotof_bufs_with_offset) {
...
+  ASSERT(r == 0);
+  uv_run(loop, UV_RUN_DEFAULT);
+  ASSERT(mkdir_cb_count == 1);
+  /* Setup Done Here */
+
+  /* Get a file descriptor for the directory */
+  r = uv_fs_open(loop,
+                 &open_req1,
+                 "test_dir",
+                 O_RDONLY | O_DIRECTORY,
```
Comment (santigimeno) [inline test/test-fs.c]: Tests are failing on some platforms. I think you should use ``UV_FS_O_RDONLY` and UV_FS_O_DIRECTORY` here.

Comment (thefourtheye) [reply test/test-fs.c]: Ah, that makes sense. Updated the PR with this change.

**Comment by santigimeno** [comment]
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/1058/

**Comment by thefourtheye** [comment]
Looks like zOS doesn't return `EISDIR` during the directory read. I couldn't find any solid evidence to reason why, but I found [this](https://www.ibm.com/support/knowledgecenter/en/SSLTBW_2.1.0/com.ibm.zos.v2r1.bpxb100/red.htm).

> There are several reasons why a read request might complete successfully with no bytes read. That is, with Return_value set to 0. For example, zero bytes are read in these cases:
> ...
> A directory is being read and the Physical File System does not support simple reads from directories. Opendir() and readdir() should be used.

Should I move zOS to the `if` branch?

**Comment by santigimeno** [comment]
> Should I move zOS to the if branch?

Probably  /cc @libuv/zos

File: test/test-fs.c [hunk 4]
```diff
@@ -3057,19 +3057,25 @@ TEST_IMPL(fs_read_dir) {
...
-    defined(__OpenBSD__) || \
-    defined(__NetBSD__)  || \
-    defined(_BSD_SOURCE) || \
-    defined(_AIX)
-  /* As of now, only BSDs and AIX support reading from a directory. */
-  ASSERT(r == 0)
+#if defined(__FreeBSD__)   || \
+    defined(__OpenBSD__)   || \
+    defined(__NetBSD__)    || \
+    defined(_BSD_SOURCE)   || \
```
Comment (carenas) [inline test/test-fs.c]: the use of this FTM here seems odd and shouldn't be needed IMHO

**Comment by richardlau** [comment]
> 
> 
> Looks like zOS doesn't return `EISDIR` during the directory read. I couldn't find any solid evidence to reason why, but I found [this](https://www.ibm.com/support/knowledgecenter/en/SSLTBW_2.1.0/com.ibm.zos.v2r1.bpxb100/red.htm).
> 
> > There are several reasons why a read request might complete successfully with no bytes read. That is, with Return_value set to 0. For example, zero bytes are read in these cases:
> > ...
> > A directory is being read and the Physical File System does not support simple reads from directories. Opendir() and readdir() should be used.
> 
> Should I move zOS to the `if` branch?

Re. z/OS there's also https://www.ibm.com/support/knowledgecenter/SSLTBW_2.2.0/com.ibm.zos.v2r2.bpxbd00/rtrea.htm which says similar:
> If the Physical File System does not support simple reads from directories, read() will return 0 if it is used for a directory. Users should use Opendir() and readdir() instead.

Which implies that there are physical file systems that _do_ support simple reads from directories.

At this point might it be simpler to flip the `#if` test for the platforms that do return `EISDIR` and have the `#else` allow `(r >= 0) || (r == UV_EISDIR)`?

**Comment by thefourtheye** [comment]
@richardlau Thanks for the pointers. I introduced a branch for Z/OS which expects the  number of bytes read to be zero.

File: test/test-fs.c (hunk 4 above)
Comment (thefourtheye) [reply test/test-fs.c]: Thanks, acknowledged.

**Comment by santigimeno** [comment]
> I introduced a branch for Z/OS which expects the number of bytes read to be zero.

I'm not sure this is correct. As I read the `zos` documentation, it'll return `0` only when the filesystem does not support directory reads, not always.

> At this point might it be simpler to flip the #if test for the platforms that do return EISDIR and have the #else allow (r >= 0) || (r == UV_EISDIR)?

I think this is a good idea. It looks like only `linux` and `windows` doesn't support reading a directory.

**Comment by thefourtheye** [comment]
@santigimeno I moved Z/OS to the lenient branch. I am not comfortable flipping the condition because, strict check will be enabled only for Windows and Linux. If we add more operating systems in the future, the check will always land in the lenient branch and we may not get a chance to review this behaviour.

**Comment by thefourtheye** [comment]
> I think this is a good idea. It looks like only linux and windows doesn't support reading a directory.

If we feel strongly about this, I can do this.

**Comment by santigimeno** [comment]
> If we feel strongly about this, I can do this.

I'm fine either way.

CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/1059/

**Comment by carenas** [comment]
> I think this is a good idea. It looks like only `linux` and `windows` doesn't support reading a directory.

macos (at least 10.14 using apfs) is also returning EISDIR and therefore passes the test through the current "else" branch

**Comment by thefourtheye** [comment]
@carenas True, MacOS passes through the strict check.

**Comment by thefourtheye** [comment]
@santigimeno Is this PR ready to land? Failures in https://ci.nodejs.org/view/libuv/job/libuv-test-commit/1059 don't look related to this PR.

**Comment by santigimeno** [comment]
Landed in https://github.com/libuv/libuv/commit/1dfa88f35b148be0afe771f1b29d251a64f32cf7. Thanks!

**Pull request closed without merging**